internal/experiment/generation/containerresources.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2020 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generation
+
+import (
+	"encoding/json"
+	"math"
+	"regexp"
+
+	redskyv1beta1 "github.com/thestormforge/optimize-controller/api/v1beta1"
+	"github.com/thestormforge/optimize-controller/internal/scan"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// ContainerResourcesSelector scans for container resources specifications (requests/limits).
+type ContainerResourcesSelector struct {
+	scan.GenericSelector
+	// Path to the list of containers.
+	Path string `json:"path,omitempty"`
+	// Create container resource specifications even if the original object does not contain them.
+	CreateIfNotPresent bool `json:"create,omitempty"`
+	// Regular expression matching the container name.
+	ContainerName string `json:"containerName,omitempty"`
+}
+
+var _ scan.Selector = &ContainerResourcesSelector{}
+
+// Map inspects the supplied resource for container resources specifications.
+func (s *ContainerResourcesSelector) Map(node *yaml.RNode, meta yaml.ResourceMeta) ([]interface{}, error) {
+	var result []interface{}
+
+	path := splitPath(s.Path) // Path to the containers list
+	err := node.PipeE(
+		yaml.Lookup(path...),
+		yaml.FilterFunc(func(node *yaml.RNode) (*yaml.RNode, error) {
+			return nil, node.VisitElements(func(node *yaml.RNode) error {
+				rl := node.Field("resources")
+				if rl == nil && !s.CreateIfNotPresent {
+					return nil
+				}
+
+				name := node.Field("name").Value.YNode().Value
+				if !s.matchesContainerName(name) {
+					return nil
+				}
+
+				p := containerResourcesParameter{pnode{
+					meta:      meta,
+					fieldPath: append(path, "[name="+name+"]", "resources"),
+				}}
+				if rl != nil {
+					p.value = rl.Value.YNode()
+				}
+
+				result = append(result, &p)
+				return nil
+			})
+		}))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// matchesContainerName checks to see if the specified container name is matched.
+func (s *ContainerResourcesSelector) matchesContainerName(name string) bool {
+	// Treat empty like ".*"
+	if s.ContainerName == "" {
+		return true
+	}
+
+	containerName, err := regexp.Compile("^(?:" + s.ContainerName + ")$")
+	if err != nil {
+		// Kustomize panics. Not sure where it validates. We will just fall back to exact match
```

Comment:
This is an old comment I probably should have cleaned up. It's in reference to the fact that Kustomize uses `MustCompile` instead of `Compile`. I never did find the code that did the error checking to give a useful error message though.